### PR TITLE
Widen mutable.BitSet operations to accept any collection.BitSet, not just mutable

### PR DESCRIPTION
**Pull request by joshlemer**
I saw no reason these methods should be restricted to only accepting mutable BitSets.
Also, replaced for expressions and Range objects with while loops. 

Also, this includes an optimized implementation of `mutable.BitSet#addAll(IterableOnce[Int])` where if  the argument is a non-negative `Range` with a step of `+/- 1`, we simply fill in all the ones from position `range.min` to `range.max`. Benchmarks below 

Note there's a big jump in cpu time at size `1000000`, this is because the BitSets are initialized to size `1000`, meaning that in the case of `1000000` an array resizing must be performed.
```
[info] Benchmark                   (size)  Mode  Cnt          Score          Error  Units
[info] BitSetBenchmark.newAddAll        0  avgt    6       3526.876 ±      189.692  ns/op
[info] BitSetBenchmark.newAddAll        3  avgt    6       3432.946 ±      105.355  ns/op
[info] BitSetBenchmark.newAddAll        5  avgt    6       3451.243 ±      205.785  ns/op
[info] BitSetBenchmark.newAddAll       10  avgt    6       3448.004 ±       91.350  ns/op
[info] BitSetBenchmark.newAddAll     1000  avgt    6       3836.286 ±       59.384  ns/op
[info] BitSetBenchmark.newAddAll  1000000  avgt    6     870262.127 ±    28773.114  ns/op
[info] BitSetBenchmark.oldAddAll        0  avgt    6       3577.063 ±       47.795  ns/op
[info] BitSetBenchmark.oldAddAll        3  avgt    6       4801.710 ±      355.115  ns/op
[info] BitSetBenchmark.oldAddAll        5  avgt    6       5516.834 ±      161.252  ns/op
[info] BitSetBenchmark.oldAddAll       10  avgt    6       7083.424 ±      118.083  ns/op
[info] BitSetBenchmark.oldAddAll     1000  avgt    6     673910.034 ±    12110.355  ns/op
[info] BitSetBenchmark.oldAddAll  1000000  avgt    6  547244039.917 ± 11428012.641  ns/op
```

benchmark code is included in this PR.

---------

- [x] Remove `oldAddAll`

### CODE REVIEW

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [inline src/library/scala/collection/mutable/BitSet.scala:163]: Note that we actually don't have to ensure capacity here, since were just removing any elements in `other` from `this`

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [inline src/library/scala/collection/mutable/BitSet.scala]: Left in for benchmarking. Should be removed before merging!

File: src/library/scala/collection/mutable/BitSet.scala
Comment (NthPortal) [inline src/library/scala/collection/mutable/BitSet.scala:194]: indentation appears to have gotten messed up here

Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala:194]: Fixed

File: src/library/scala/collection/mutable/BitSet.scala
Comment (NthPortal) [inline src/library/scala/collection/mutable/BitSet.scala:197]: this appears to duplicate the guard in the `case` clause

Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala:197]: Ah yes thanks! I recently brought the guard down out of the case statement so that even if the step is not by `+- 1`, we can still take the opportunity to perform an `ensureCapacity` by performing the fast `range.max` operation, and then falling back to `super.addAll`

File: src/library/scala/collection/mutable/BitSet.scala
Comment (NthPortal) [inline src/library/scala/collection/mutable/BitSet.scala]: this can go out of bounds of `bs`

**Review by NthPortal — COMMENTED** [review]
This looks pretty good to me, but I don't consider myself well-versed enough with the implementation to give an approval.

**Comment by joshlemer** [comment]
The failing test looks unrelated to this change, not sure how to restart jenkins..

File: src/library/scala/collection/mutable/BitSet.scala
Comment (Ichoran) [inline src/library/scala/collection/mutable/BitSet.scala]: Could you break this out onto a separate line?  It's hard to notice it's even a case as a one-liner!

File: src/library/scala/collection/mutable/BitSet.scala
Comment (Ichoran) [inline src/library/scala/collection/mutable/BitSet.scala]: Could you one-liner-ize all these `else`s?  It's really hard to follow the nesting depth with all the ins-and-outs.  So

```
if (p) {
  stuff
  more.stuff
} else {
  thing
}
```

becomes

```
if (p) {
  stuff
  more.stuff
}
else thing
```

I know that having braces everywhere is said by some to be good practice, but I think it really interferes with code clarity here.

File: src/library/scala/collection/mutable/BitSet.scala
Comment (Ichoran) [inline src/library/scala/collection/mutable/BitSet.scala]: Are we sure this is necessary?  It's certainly possible, but is `PriorityQueue[Int]` really used enough to be worth the effort?

File: src/library/scala/collection/mutable/BitSet.scala
Comment (Ichoran) [inline src/library/scala/collection/mutable/BitSet.scala]: I think the logic could be simplified here as follows.  First, if `thiswords > bsnwords`, we check all the high ones and return `bsnwords` as the remainder to check; otherwise we do nothing and use `thiswords` as the number to check.  Second, check from 0 up to that number.  This way you won't have the code duplication of checking the low-index entries in two places.

**Review by Ichoran — CHANGES_REQUESTED** [review]
Some formatting changes, but the logic mostly looks good to me.  (Assuming that refactoring to reduce code duplication counts as a "formatting change"?)

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala]: I think maybe for pure code this is more acceptable, but in side-effecting blocks, delimiting with curly braces helps protect against error where a second statement is assumed to be in the else branch but actually is outside.

**Comment by SethTisue** [comment]
needs rebase

**Comment by Ichoran** [comment]
@joshlemer - Do you have time to rebase and address the remaining comments?  The awkward-to-me code style is acceptable, though I would personally at least make the one-liners one-liners in braces.  The other stuff needs a look, though.

File: src/library/scala/collection/mutable/BitSet.scala
Comment (szeiger) [inline src/library/scala/collection/mutable/BitSet.scala]: A single `this` at the end would suffice. The method returns `this.type` so the branches going through other methods such as `super.addAll` can't return anything else.

File: src/library/scala/collection/mutable/BitSet.scala
Comment (szeiger) [inline src/library/scala/collection/mutable/BitSet.scala]: Manually iterating through the `Range` would probably be much faster than using the standard `Growable.addAll` implementation (mainly to avoid boxing) but it doesn't have to be part of this PR.

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala]: I have removed this code now

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala]: Consider it simplified!

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala]: Done!

File: src/library/scala/collection/mutable/BitSet.scala
Comment (joshlemer) [reply src/library/scala/collection/mutable/BitSet.scala]: I think I'll leave that to a later, the Range#foreach code can't be directly inlined here because it accesses `private[this] val lastElement`

**Comment by joshlemer** [comment]
@Ichoran I think this is ready for re-review/approval. I think all the braces you didn't like have been turned into one-liners

**Review by szeiger — APPROVED** [review]

**Pull request merged**
